### ISSUE TITLE
Using cls Argument of Classmethod Defined Before __init__ Causes Error

### ISSUE THREAD

**Issue by jannero** [python/mypy#5415]
Mypy is giving an error `Too many arguments for "<class>"` if _cls_ argument is used for creating object in a classmethod that is defined before _\_\_init\_\__. I would expect this to work as similar constructs (see below) work.

The case for having separate class method _create_ is to allow doing some initializations in certain cases (when the classmethod is called) but not in others (when the _\_\_init\_\__ is called). The example code is not complete in that sense but is enough to demonstrate the issue I faced.

Used versions:
* Mypy: 0.620 and 0.630+dev-132dfa264e983a0c566f3fa6f316cde27afff131, i.e. the latest version from _master_ branch
* Python: 3.6 and 3.7 (haven't tried others)

The output from `mypy example.py`: `example.py:4: error: Too many arguments for "A"` for the code below:

```python
class A:
    @classmethod
    def create(cls) -> A:
        return cls(0)

    def __init__(self, arg: int) -> None:
        self.arg = arg
```

The code below contains similar constructs (in addition to the case above) that do not cause mypy to report errors, i.e. the only reported error is from _create1_ class method (and the error is the same as in the above code):

```python
def foo() -> A:
    return bar(9)


def bar(value: int) -> A:
    return A(value)


class A:
    # The only error is reported from this class method
    @classmethod
    def create1(cls) -> A:
        return cls(0)

    @classmethod
    def create2(cls) -> A:
        return A(0)

    @staticmethod
    def create3() -> A:
        return A(1)

    def __init__(self, arg: int) -> None:
        self.arg = arg

    @classmethod
    def create4(cls) -> A:
        return cls(0)

    @classmethod
    def create5(cls) -> A:
        return A(0)

    @staticmethod
    def create6() -> A:
        return A(1)
```
